### ISSUE TITLE
Can't build opentsdb 2.2.0

### ISSUE THREAD

**Issue by CamJN** [OpenTSDB/opentsdb#707]
Grabbed source tarball from https://github.com/OpenTSDB/opentsdb/releases/download/v2.2.0/opentsdb-2.2.0.tar.gz and extracted, and went to that dir.

There's lots of output when I try and build opentsdb, but I think the most relevant error is: `No rule to make target 'net/opentsdb/core/AggregationIterator.class', needed by 'tsdb-2.2.0.jar'.  Stop.`

Full output
```
bash-4.3$ ./build.sh 
+ test -f configure
+ test -d build
+ mkdir build
+ cd build
+ test -f Makefile
+ ../configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ../build-aux/install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for md5sum... no
checking for md5... /sbin/md5
checking for java... /Library/Java/JavaVirtualMachines/jdk1.8.0_72.jdk/Contents/Home/bin/java
checking for javac... /Library/Java/JavaVirtualMachines/jdk1.8.0_72.jdk/Contents/Home/bin/javac
checking for jar... /Library/Java/JavaVirtualMachines/jdk1.8.0_72.jdk/Contents/Home/bin/jar
checking for true... /usr/bin/true
checking for javadoc... /Library/Java/JavaVirtualMachines/jdk1.8.0_72.jdk/Contents/Home/bin/javadoc
checking for wget... no
checking for curl... /usr/bin/curl
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating opentsdb.spec
config.status: creating build-aux/fetchdep.sh
+ MAKE=make
++ uname -s
+ '[' Darwin = FreeBSD ']'
+ exec make
../build-aux/gen_build_data.sh src/tools/BuildData.java net.opentsdb.tools 2.2.0
Generating src/tools/BuildData.java
fatal: Not a git repository (or any of the parent directories): .git
fatal: Not a git repository (or any of the parent directories): .git
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-am
rm -f tsdb tsdb.tmp
script=tsdb; pkgdatadir=''; \
          abs_srcdir='/private/tmp/opentsdb20160224-99228-12y38mp/opentsdb-2.2.0/build/..'; abs_builddir='/private/tmp/opentsdb20160224-99228-12y38mp/opentsdb-2.2.0/build'; \
          srcdir=''; test -f ./$script.in || srcdir=../; sed -e "s:@pkgdatadir[@]:$pkgdatadir:g" -e "s:@abs_srcdir[@]:$abs_srcdir:g" -e "s:@abs_builddir[@]:$abs_builddir:g" -e "s:@configdir[@]:$configdir:g" ${srcdir}$script.in >$script.tmp
make[1]: *** No rule to make target `net/opentsdb/core/AggregationIterator.class', needed by `tsdb-2.2.0.jar'.  Stop.
make[1]: *** Waiting for unfinished jobs....
chmod +x tsdb.tmp
chmod a-w tsdb.tmp
mv tsdb.tmp tsdb
/Library/Java/JavaVirtualMachines/jdk1.8.0_72.jdk/Contents/Home/bin/javac -Xlint -source 6 -encoding utf-8 -d . -cp ../third_party/hbase/asynchbase-1.7.1.jar:../third_party/guava/guava-18.0.jar:../third_party/slf4j/log4j-over-slf4j-1.7.7.jar:../third_party/logback/logback-classic-1.0.13.jar:../third_party/logback/logback-core-1.0.13.jar:../third_party/jackson/jackson-annotations-2.4.3.jar:../third_party/jackson/jackson-core-2.4.3.jar:../third_party/jackson/jackson-databind-2.4.3.jar:../third_party/netty/netty-3.9.4.Final.jar:../third_party/protobuf/protobuf-java-2.5.0.jar:../third_party/slf4j/slf4j-api-1.7.7.jar:../third_party/suasync/async-1.4.0.jar:../third_party/zookeeper/zookeeper-3.4.5.jar:../third_party/apache/commons-math3-3.4.1.jar:  ../src/core/AggregationIterator.java ../src/core/Aggregator.java ../src/core/Aggregators.java ../src/core/AppendDataPoints.java ../src/core/BatchedDataPoints.java ../src/core/ByteBufferList.java ../src/core/ColumnDatapointIterator.java ../src/core/CompactionQueue.java ../src/core/Const.java ../src/core/DataPoint.java ../src/core/DataPoints.java ../src/core/DataPointsIterator.java ../src/core/Downsampler.java ../src/core/DownsamplingSpecification.java ../src/core/FillingDownsampler.java ../src/core/FillPolicy.java ../src/core/IncomingDataPoint.java ../src/core/IncomingDataPoints.java ../src/core/IllegalDataException.java ../src/core/Internal.java ../src/core/MutableDataPoint.java ../src/core/Query.java ../src/core/QueryException.java ../src/core/RateOptions.java ../src/core/RateSpan.java ../src/core/RowKey.java ../src/core/RowSeq.java ../src/core/SaltScanner.java ../src/core/SeekableView.java ../src/core/Span.java ../src/core/SpanGroup.java ../src/core/TSDB.java ../src/core/Tags.java ../src/core/TsdbQuery.java ../src/core/TSQuery.java ../src/core/TSSubQuery.java ../src/core/WritableDataPoints.java ../src/graph/Plot.java ../src/meta/Annotation.java ../src/meta/TSMeta.java ../src/meta/TSUIDQuery.java ../src/meta/UIDMeta.java ../src/query/QueryUtil.java ../src/query/filter/TagVFilter.java ../src/query/filter/TagVLiteralOrFilter.java ../src/query/filter/TagVNotKeyFilter.java ../src/query/filter/TagVNotLiteralOrFilter.java ../src/query/filter/TagVRegexFilter.java ../src/query/filter/TagVWildcardFilter.java ../src/search/SearchPlugin.java ../src/search/SearchQuery.java ../src/search/TimeSeriesLookup.java ../src/stats/Histogram.java ../src/stats/StatsCollector.java ../src/stats/QueryStats.java ../src/tools/ArgP.java ../src/tools/CliOptions.java ../src/tools/CliQuery.java ../src/tools/CliUtils.java ../src/tools/DumpSeries.java ../src/tools/Fsck.java ../src/tools/FsckOptions.java ../src/tools/MetaPurge.java ../src/tools/MetaSync.java ../src/tools/Search.java ../src/tools/TSDMain.java ../src/tools/TextImporter.java ../src/tools/TreeSync.java ../src/tools/UidManager.java ../src/tree/Branch.java ../src/tree/Leaf.java ../src/tree/Tree.java ../src/tree/TreeBuilder.java ../src/tree/TreeRule.java ../src/tsd/AbstractHttpQuery.java ../src/tsd/AnnotationRpc.java ../src/tsd/BadRequestException.java ../src/tsd/ConnectionManager.java ../src/tsd/GnuplotException.java ../src/tsd/GraphHandler.java ../src/tsd/HttpJsonSerializer.java ../src/tsd/HttpSerializer.java ../src/tsd/HttpQuery.java ../src/tsd/HttpRpc.java ../src/tsd/HttpRpcPlugin.java ../src/tsd/HttpRpcPluginQuery.java ../src/tsd/LineBasedFrameDecoder.java ../src/tsd/LogsRpc.java ../src/tsd/PipelineFactory.java ../src/tsd/PutDataPointRpc.java ../src/tsd/QueryRpc.java ../src/tsd/RpcHandler.java ../src/tsd/RpcPlugin.java ../src/tsd/RpcManager.java ../src/tsd/RTPublisher.java ../src/tsd/SearchRpc.java ../src/tsd/StaticFileRpc.java ../src/tsd/StatsRpc.java ../src/tsd/StorageExceptionHandler.java ../src/tsd/SuggestRpc.java ../src/tsd/TelnetRpc.java ../src/tsd/TreeRpc.java ../src/tsd/UniqueIdRpc.java ../src/tsd/WordSplitter.java ../src/uid/FailedToAssignUniqueIdException.java ../src/uid/NoSuchUniqueId.java ../src/uid/NoSuchUniqueName.java ../src/uid/RandomUniqueId.java ../src/uid/UniqueId.java ../src/uid/UniqueIdInterface.java ../src/utils/ByteArrayPair.java ../src/utils/Config.java ../src/utils/DateTime.java ../src/utils/Exceptions.java ../src/utils/FileSystem.java ../src/utils/JSON.java ../src/utils/JSONException.java ../src/utils/Pair.java ../src/utils/PluginLoader.java ../src/utils/Threads.java ../src/tools/BuildData.java
warning: [options] bootstrap class path not set in conjunction with -source 1.6
../src/uid/UniqueId.java:197: warning: [dep-ann] deprecated item is not annotated with @Deprecated
  public long maxPossibleId() {
              ^
../src/uid/UniqueId.java:441: warning: [rawtypes] found raw type: Deferred
      final Deferred d;
            ^
  missing type arguments for generic class Deferred<T>
  where T is a type-variable:
    T extends Object declared in class Deferred
../src/core/TSSubQuery.java:237: warning: [dep-ann] deprecated item is not annotated with @Deprecated
  public Map<String, String> getTags() {
                             ^
../src/core/TSSubQuery.java:295: warning: [dep-ann] deprecated item is not annotated with @Deprecated
  public void setTags(Map<String, String> tags) {
              ^
../src/meta/TSUIDQuery.java:92: warning: [dep-ann] deprecated item is not annotated with @Deprecated
  public TSUIDQuery(final TSDB tsdb) {
         ^
../src/meta/TSUIDQuery.java:226: warning: [dep-ann] deprecated item is not annotated with @Deprecated
  public void setQuery(final String metric, final Map<String, String> tags) {
              ^
../src/meta/TSUIDQuery.java:550: warning: [dep-ann] deprecated item is not annotated with @Deprecated
  public static Deferred<IncomingDataPoint> getLastPoint(final TSDB tsdb, 
                                            ^
../src/tools/Fsck.java:809: warning: [cast] redundant cast to float
                Float.floatToRawIntBits((float)value_as_float));
                                        ^
../src/tools/Fsck.java:853: warning: [cast] redundant cast to float
                Float.floatToRawIntBits((float)value_as_float));
                                        ^
../src/tsd/AbstractHttpQuery.java:129: warning: [deprecation] getHeaders() in HttpMessage has been deprecated
        request.getHeaders().size());
               ^
../src/tsd/AbstractHttpQuery.java:130: warning: [deprecation] getHeaders() in HttpMessage has been deprecated
    for (final Entry<String, String> header : request.getHeaders()) {
                                                     ^
../src/tsd/AbstractHttpQuery.java:155: warning: [deprecation] getHeaders() in HttpMessage has been deprecated
        request.getHeaders().size());
               ^
../src/tsd/AbstractHttpQuery.java:156: warning: [deprecation] getHeaders() in HttpMessage has been deprecated
    for (final Entry<String, String> header : request.getHeaders()) {
                                                     ^
../src/tsd/AbstractHttpQuery.java:451: warning: [deprecation] toStringHelper(Object) in Objects has been deprecated
    return Objects.toStringHelper(this)
                  ^
../src/tsd/QueryRpc.java:374: warning: [deprecation] getLastPoint(TSDB,byte[],boolean,int,long) in TSUIDQuery has been deprecated
          deferreds.add(TSUIDQuery.getLastPoint(tsdb, entry.getKey(), 
                                  ^
../src/tsd/GraphHandler.java:188: warning: [rawtypes] found raw type: HashSet
    final HashSet<String>[] aggregated_tags = new HashSet[nqueries];
                                                  ^
  missing type arguments for generic class HashSet<E>
  where E is a type-variable:
    E extends Object declared in class HashSet
../src/tsd/LogsRpc.java:115: warning: [cast] redundant cast to ILoggingEvent
        final ILoggingEvent event = (ILoggingEvent) logbuf.get(nevents);
                                    ^
../src/utils/JSON.java:214: warning: [deprecation] createJsonParser(String) in JsonFactory has been deprecated
      return jsonMapper.getFactory().createJsonParser(json);
                                    ^
../src/utils/JSON.java:236: warning: [deprecation] createJsonParser(byte[]) in JsonFactory has been deprecated
      return jsonMapper.getFactory().createJsonParser(json);
                                    ^
../src/utils/JSON.java:258: warning: [deprecation] createJsonParser(InputStream) in JsonFactory has been deprecated
      return jsonMapper.getFactory().createJsonParser(json);
                                    ^
../src/utils/PluginLoader.java:69: warning: [rawtypes] found raw type: Class
  private static final Class<?>[] PARAMETER_TYPES = new Class[] {
                                                        ^
  missing type arguments for generic class Class<T>
  where T is a type-variable:
    T extends Object declared in class Class
22 warnings
make: *** [all] Error 2
```

**Comment by CamJN** [comment]
Something odd is that `build/net/opentsdb/core/AggregationIterator.class` exists. So obviously it was built...

**Comment by CamJN** [comment]
removing `$(classes)` from https://github.com/OpenTSDB/opentsdb/blob/master/Makefile.am#L614 worked to compile the jar, will start testing if it actually works now.

**Comment by CamJN** [comment]
at this point `make install-data-am` is failing because at some point it tries to call `../build-aux/install-sh -c -d '/usr/local/Cellar/opentsdb/2.2.0/share/opentsdb/static/gwt/opentsdb/images'` outside of the build dir and fails.

**Comment by CamJN** [comment]
that seems to be because of the `cd` at https://github.com/OpenTSDB/opentsdb/blob/master/Makefile.am#L455 combined with the `mkdir_p` at https://github.com/OpenTSDB/opentsdb/blob/master/Makefile.am#L461

Since this script seems to be somewhat neglected, I'm wondering what the maintained method of building opentsdb is? Because piecing together a functioning build from the current state of the makefile and meta-makefile boilerplate has proven difficult.

**Comment by manolama** [comment]
Hiya, what problem do you run into? I just tried downloading the tar to a VM, built it with ``sh build.sh`` and installed with ``sudo sh build.sh install`` and it worked properly. Were you trying to run the makefile directly?

**Comment by CamJN** [comment]
At first I tried running `build.sh` but `make` couldn't get through the `Makefile`. Then I tried running the various `make` jobs individually in the order that they appear in the `Makefile` to see where things failed. And then I had to reorder them a bit and use the changes I described in my patch.

The order I had to do things in so that files would be where they were expected was:
```
make
make install-exec-am
make install-data-am
```
I'm still seeing some weird behavior when I try and start `hbase shell` until I open a fresh terminal:
```
NameError: uninitialized constant Formula
  const_missing at org/jruby/RubyModule.java:2647
         (root) at ./readline.rb:1
        require at org/jruby/RubyKernel.java:1062
         (root) at ./readline.rb:10
        require at org/jruby/RubyKernel.java:1062
         (root) at /usr/local/opt/hbase/libexec/bin/../bin/hirb.rb:35
```

**Comment by CamJN** [comment]
Specifically I'm creating a homebrew formula for opentsdb and those were the issues I've run into so far.

You can see it at https://github.com/CamJN/homebrew/blob/opentsdb/Library/Formula/opentsdb.rb though it's not done (waiting on some hbase formula changes to be made).

**Comment by manolama** [comment]
ah ok, on a mac then :)  Hmm, which version of make do you use? And thanks for the brew formula! That'd be awesome!

**Comment by CamJN** [comment]
It's GNU Make 3.81 built for i386-apple-darwin11.3.0

**Comment by johann8384** [comment]
In case it is useful, this works for me: https://github.com/OpenTSDB/opentsdb/blob/next/tools/osx_full_stack_install.sh

**Comment by johann8384** [comment]
I'll test the tarball on my mac. I have never tried to build the tarball, but I have built from the repo many times.

**Comment by johann8384** [comment]
Compiles and runs for me, @CamJN are you still having this problem?

**Comment by CamJN** [comment]
I just noticed that @manolama used `sh build.sh` not `build.sh`, and when i tried to build that way it seems to have worked.

So I have to ask, why run a bash script from sh? It just spawns a bash subshell anyway. It does seem to work, but I don't see why it would make a difference.

**Comment by johann8384** [comment]
No idea, and I just do "./build.sh" usually, and I seem to have the same
env as you.

**Comment by CamJN** [comment]
Just being on OS X doesn't mean you have the same env, just the same platform, some examples: 

- there are 19 ENV vars defined in bash on os x by default, I have 60; 
- there are 4 paths in PATH by default, I have 19; 
- there are 0 aliases by default, I have 76;
- there are 8 functions defined by default, I have 448;
- os x includes bash 3.2.57 and I use 4.3.42;

however homebrew has done the work to standardize the env in their build process, so it shouldn't matter what my normal env is like; just what they set up. You can see more here: https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/extend/ENV/super.rb

**Comment by johann8384** [comment]
Yes of course. My env is far from default, but I meant to indicate that it
wasn't as different as you being on Linux and me being on AIX or something.
:)

-Jonathan

**Comment by johann8384** [comment]
Obviously there are differences or we wouldn't be having this conversation.
:)

So, at the moment I'm downloading the release tarball again and I was going
to try a few incantations of the command to see if I can break it.

**Comment by manolama** [comment]
This one may need some playing around with. I just found that it broke the RPM build when I was trying to deploy 2.3.0RC1. :(

**Comment by CamJN** [comment]
New homebrew formula link (because of the refactoring they did): https://github.com/Homebrew/homebrew-core/pull/813

**Comment by CamJN** [comment]
Formula was accepted into homebrew, so you can review the final form at https://github.com/Homebrew/homebrew-core/blob/master/Formula/opentsdb.rb

**Comment by CheRuisiBesares** [comment]
Looks like this is getting the attention that it needs just wanted to say that this is effecting the build on alpine linux and therefor the various docker images. It seems for now people are working around this by just running everything out of the build dir and not doing the install.

**Comment by johann8384** [comment]
This seems to be resolved, CamJN has completed the Homebrew formula and it appears to be working fine from my testing on two macs.
